pkg/apiserver/profiling/profile.go

```diff
@@ -15,245 +15,22 @@ package profiling
-	f, err := os.Create(o.output)
-	return f, err
-}
-
-type fetcher struct {
-	ctx        context.Context
-	httpClient *httpc.Client
-	target     *model.RequestTargetNode
-	output     string
-	httpScheme string
-}
-
-func (f *fetcher) Fetch(src string, duration, timeout time.Duration) (*profile.Profile, string, error) {
-	var url string
-	secs := strconv.Itoa(int(duration / time.Second))
-	switch f.target.Kind {
-	case model.NodeKindPD:
-		url = "/pd/api/v1/debug/pprof/profile?seconds=" + secs
-	case model.NodeKindTiDB:
-		url = "/debug/pprof/profile?seconds=" + secs
-	default:
-		return nil, "", fmt.Errorf("unsupported target %s", f.target)
-	}
-	url = fmt.Sprintf("%s://%s:%d%s", f.httpScheme, f.target.IP, f.target.Port, url)
-
-	p, err := f.getProfile(f.target, url)
-	return p, url, err
-}
-
-func (f *fetcher) getProfile(target *model.RequestTargetNode, source string) (*profile.Profile, error) {
-	req, err := http.NewRequestWithContext(f.ctx, http.MethodGet, source, nil)
-	if err != nil {
-		return nil, fmt.Errorf("failed to create a new request %s: %v", source, err)
-	}
-	if target.Kind == model.NodeKindPD {
-		// forbidden PD follower proxy
-		req.Header.Add("PD-Allow-follower-handle", "true")
-	}
-	resp, err := f.httpClient.WithTimeout(maxProfilingTimeout).Do(req)
-	if err != nil {
-		return nil, fmt.Errorf("request %s failed: %v", source, err)
-	}
-	defer resp.Body.Close()
-	if resp.StatusCode != http.StatusOK {
-		return nil, fmt.Errorf("failed to profile %s: status code %s", target, resp.Status)
-	}
-	return profile.Parse(resp.Body)
-}
-
-func profileAndWriteSVG(ctx context.Context, target *model.RequestTargetNode, fileNameWithoutExt string, profileDurationSecs uint, httpClient *httpc.Client, httpScheme string) (string, error) {
+func profileAndWriteSVG(ctx context.Context, fts *fetchers, target *model.RequestTargetNode, fileNameWithoutExt string, profileDurationSecs uint) (string, error) {
 	switch target.Kind {
-	case model.NodeKindTiKV, model.NodeKindTiFlash:
-		return fetchFlameGraphSVG(ctx, httpClient, target, fileNameWithoutExt, profileDurationSecs, httpScheme)
-	case model.NodeKindPD, model.NodeKindTiDB:
-		return fetchPprofSVG(ctx, httpClient, target, fileNameWithoutExt, profileDurationSecs, httpScheme)
+	case model.NodeKindTiKV:
+		return fetchFlameGraphSVG(&flameGraphOptions{duration: profileDurationSecs, fileNameWithoutExt: fileNameWithoutExt, target: target, fetcher: &fts.tikv})
+	case model.NodeKindTiFlash:
+		return fetchFlameGraphSVG(&flameGraphOptions{duration: profileDurationSecs, fileNameWithoutExt: fileNameWithoutExt, target: target, fetcher: &fts.tiflash})
+	case model.NodeKindTiDB:
+		return fetchPprofSVG(&pprofOptions{duration: profileDurationSecs, fileNameWithoutExt: fileNameWithoutExt, target: target, fetcher: &fts.tidb})
+	case model.NodeKindPD:
+		return fetchPprofSVG(&pprofOptions{duration: profileDurationSecs, fileNameWithoutExt: fileNameWithoutExt, target: target, fetcher: &fts.pd})
```

Comment:
I think you can further refine this part. Now you have separated different fetchers, but after fetching different components are still sticked together, so that it is still hard to implement a new profiling component, require modifying codes everywhere. Also maybe you can consider this feature requirement into consideration when doing the refactoring: https://github.com/pingcap/tidb-dashboard/issues/717

Reply:
Yes, I'm considering the new feature `Debugging API UI` in the `Spec: View Server Logs and Status` doc, and will continue to refine this part.